core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java

```diff
@@ -185,33 +180,30 @@ private static HttpHeaders getHttpHeaders(HttpResponse res) {
         return future.join();
     }
 
-    private void retry0(int currentAttemptNo, Backoff backoff, ClientRequestContext ctx,
-                        HttpRequestDuplicator rootReqDuplicator, Function<HttpRequest, HttpResponse> action,
-                        DeferredHttpResponse deferredRes, Throwable exception, long millisAfter) {
-        long nextDelay = backoff.nextDelayMillis(currentAttemptNo);
-        if (nextDelay < 0) {
-            deferredRes.close(exception);
-            rootReqDuplicator.close();
-        } else {
-            final EventLoop eventLoop = ctx.contextAwareEventLoop();
-            nextDelay = Math.max(nextDelay, millisAfter);
-            try {
-                nextDelay = getNextDelay(nextDelay, ctx);
-            } catch (ResponseTimeoutException e) {
-                deferredRes.close(e);
+    private void retry0(ClientRequestContext ctx, Backoff backoff, HttpRequestDuplicator rootReqDuplicator,
+                        Function<HttpRequest, HttpResponse> action, DeferredHttpResponse deferredRes,
+                        long millisAfter) {
+        long nextDelay;
+        try {
+            nextDelay = getNextDelay(ctx, backoff, millisAfter);
+            if (nextDelay < 0) { // exceed the number of max attempt
+                deferredRes.close(RetryGiveUpException.get());
                 rootReqDuplicator.close();
                 return;
             }
-            if (nextDelay <= 0) {
-                eventLoop.submit(() -> retry(currentAttemptNo + 1, backoff, ctx,
-                                             rootReqDuplicator, action,
-                                             deferredRes));
-            } else {
-                eventLoop.schedule(
-                        () -> retry(currentAttemptNo + 1, backoff, ctx,
-                                    rootReqDuplicator, action, deferredRes),
-                        nextDelay, TimeUnit.MILLISECONDS);
-            }
+        } catch (ResponseTimeoutException e) {
+            deferredRes.close(e);
+            rootReqDuplicator.close();
+            return;
+        }
+
+        final EventLoop eventLoop = ctx.contextAwareEventLoop();
+        if (nextDelay <= 0) {
```

Comment:
In this case, `nextDelay` is always `0`, right? If so, how about rearranging the codes like below for readability?
```
final long nextDelay = getNextDelay(...);
if (nextDelay == 0) {
  eventLoop.submit(...);
  return;
} 
if (nextDelay > 0) {
  eventLoop.schedule(...);
  return;
}
try {
...
} catch (ResponseTimeoutException e) {
...
}
```